### PR TITLE
Handle IForToLoopOperation over object in control flow graph

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review.

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb
Comment (mavasani) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: I presume this is the same anonymous local that is declared in the enclosing region `R1`? Wondering why the name for this is blank, while the dump above contains `<anonymous local>`

**Review by mavasani — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Nevermind, I see that it's out of order.

---
In reply to: [188116692](https://github.com/dotnet/roslyn/pull/26803#discussion_r188116692) [](ancestors = 188116692)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >mysterious [](start = 46, length = 10)

Probably not the right word :)

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Probably not the right word :)

Works for VB lowering, where I copied this comment from.

---
In reply to: [188118434](https://github.com/dotnet/roslyn/pull/26803#discussion_r188118434) [](ancestors = 188118434)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: > I presume this is the same anonymous local that is declared in the enclosing region R1? Wondering why the name for this is blank, while the dump above contains <anonymous local>

This is a difference in how the tree serialization and graph serialization are implemented
OperationTreeVerifier.VisitLocalReference:
```
LogString($": {operation.Local.Name}");
```

ControlFlowGraphVerifier.printLocals:
```
stringBuilder.Append($" [{local.ToTestDisplayString()}]");
```

A change to OperationTreeVerifier.VisitLocalReference is going to "break" a lot of tests.

---
In reply to: [188112480](https://github.com/dotnet/roslyn/pull/26803#discussion_r188112480) [](ancestors = 188112480)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Can the code in this and checkLoopCondition() be refactored together? It seems like the only big differences is how many things are put onto the eval stack, which could be controlled by a flag?

**Review by 333fred — APPROVED** [review]
LGTM (Iteration 2) - Comments can be addressed in a follow up.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Can the code in this and checkLoopCondition() be refactored together? It seems like the only big differences is how many things are put onto the eval stack, which could be controlled by a flag?

I'll se what I can do (in a follow-up PR)

---
In reply to: [188128275](https://github.com/dotnet/roslyn/pull/26803#discussion_r188128275) [](ancestors = 188128275)

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review, need a second review.

**Review by cston — APPROVED** [review]

**Pull request merged**
